Get StackOverflow issue for ballerina to openapi command run

**Description:**
Given ballerina file include cyclic record as their attribute type. due to this cyclic record, StackOverflow occurs
```ballerina
public type InnerError record {
    # Error code.
    string code;
    # Error message.
    string message;
    # Error details.
    record {} details?;
    # Error target.
    string target?;
    InnerError innererror?;
};
```

stack trace
```
[2022-01-20 09:54:55,346] SEVERE {b7a.log.crash} - The compiler extension 'io.ballerina.openapi.extension.OpenApiCompilerPlugin' failed to complete. null 
io.ballerina.projects.ProjectException: The compiler extension 'io.ballerina.openapi.extension.OpenApiCompilerPlugin' failed to complete. null
	at io.ballerina.projects.SyntaxNodeAnalysisTask.perform(SyntaxNodeAnalysisTask.java:62)
	at io.ballerina.projects.SyntaxTreeVisitor.runAnalysisTasks(SyntaxTreeVisitor.java:93)
	at io.ballerina.projects.SyntaxTreeVisitor.visitSyntaxNode(SyntaxTreeVisitor.java:80)
	at io.ballerina.compiler.syntax.tree.NodeVisitor.visit(NodeVisitor.java:63)
	at io.ballerina.compiler.syntax.tree.ServiceDeclarationNode.accept(ServiceDeclarationNode.java:78)
	at io.ballerina.projects.SyntaxTreeVisitor.visitSyntaxNode(SyntaxTreeVisitor.java:85)
	at io.ballerina.compiler.syntax.tree.NodeVisitor.visit(NodeVisitor.java:43)
	at io.ballerina.projects.SyntaxTreeVisitor.runAnalysisTasks(SyntaxTreeVisitor.java:68)
	at io.ballerina.projects.SyntaxNodeAnalysisTaskRunner.runTasks(SyntaxNodeAnalysisTaskRunner.java:70)
	at io.ballerina.projects.SyntaxNodeAnalysisTaskRunner.runTasks(SyntaxNodeAnalysisTaskRunner.java:59)
	at io.ballerina.projects.SyntaxNodeAnalysisTaskRunner.runTasks(SyntaxNodeAnalysisTaskRunner.java:51)
	at io.ballerina.projects.CodeAnalyzerManager.runSyntaxNodeAnalysisTasks(CodeAnalyzerManager.java:108)
	at io.ballerina.projects.CodeAnalyzerManager.runCodeAnalyzerTasks(CodeAnalyzerManager.java:61)
	at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:129)
	at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:103)
	at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:214)
	at io.ballerina.projects.Package.getCompilation(Package.java:141)
	at io.ballerina.openapi.generators.openapi.OpenApiConverter.generateOAS3DefinitionsAllService(OpenApiConverter.java:98)
	at io.ballerina.openapi.cmd.OpenApiCmd.ballerinaToOpenApi(OpenApiCmd.java:193)
	at io.ballerina.openapi.cmd.OpenApiCmd.execute(OpenApiCmd.java:157)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:51)
Caused by: java.lang.StackOverflowError
	at io.ballerina.compiler.api.impl.symbols.TypesFactory.getTypeDescriptor(TypesFactory.java:164)
	at io.ballerina.compiler.api.impl.symbols.TypesFactory.getTypeDescriptor(TypesFactory.java:154)
	at io.ballerina.compiler.api.impl.SymbolFactory.createTypeDefinition(SymbolFactory.java:421)
	at io.ballerina.compiler.api.impl.SymbolFactory.getBCompiledSymbol(SymbolFactory.java:200)
	at io.ballerina.compiler.api.impl.symbols.BallerinaTypeReferenceTypeSymbol.definition(BallerinaTypeReferenceTypeSymbol.java:97)
```
**Steps to reproduce:**

**Affected Versions:**
swanLake beta06
**OS, DB, other environment details and versions:**

**Related Issues (optional):**
https://github.com/wso2-enterprise/choreo/issues/10791

**Suggested Labels (optional):**
<!-- Optional comma separated list of suggested labels. Non committers can’t assign labels to issues, so this will help issue creators who are not a committer to suggest possible labels-->

**Suggested Assignees (optional):**
<!--Optional comma separated list of suggested team members who should attend the issue. Non committers can’t assign issues to assignees, so this will help issue creators who are not a committer to suggest possible assignees-->

--- a/openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java
+++ b/openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java
@@ -184,7 +184,8 @@ public class OpenAPIComponentMapper {
                         typeReference));
                 schema = components.getSchemas();
             } else if (field.getValue().typeDescriptor().typeKind() == TypeDescKind.UNION) {
-                property = handleUnionType((UnionTypeSymbol) field.getValue().typeDescriptor(), property, componentName);
+                property = handleUnionType((UnionTypeSymbol) field.getValue().typeDescriptor(), property,
+                        componentName);
                 schema = components.getSchemas();
             }
             if (property instanceof ArraySchema) {
@@ -266,6 +267,9 @@ public class OpenAPIComponentMapper {
     }
 
     private boolean isSameRecord(String parentComponentName, TypeReferenceTypeSymbol typeReferenceTypeSymbol) {
+        if (parentComponentName == null) {
+            return false;
+        }
         return typeReferenceTypeSymbol.getName().isPresent() &&
                 parentComponentName.equals(typeReferenceTypeSymbol.getName().get().trim());
     }

--- a/openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java
+++ b/openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java
@@ -180,10 +180,11 @@ public class OpenAPIComponentMapper {
             Schema property = ConverterCommonUtils.getOpenApiSchema(type);
             if (field.getValue().typeDescriptor().typeKind() == TypeDescKind.TYPE_REFERENCE) {
                 TypeReferenceTypeSymbol typeReference = (TypeReferenceTypeSymbol) field.getValue().typeDescriptor();
-                property = handleTypeReference(schema, typeReference, property);
+                property = handleTypeReference(schema, typeReference, property, isSameRecord(componentName,
+                        typeReference));
                 schema = components.getSchemas();
             } else if (field.getValue().typeDescriptor().typeKind() == TypeDescKind.UNION) {
-                property = handleUnionType((UnionTypeSymbol) field.getValue().typeDescriptor(), property);
+                property = handleUnionType((UnionTypeSymbol) field.getValue().typeDescriptor(), property, componentName);
                 schema = components.getSchemas();
             }
             if (property instanceof ArraySchema) {
@@ -214,13 +215,15 @@ public class OpenAPIComponentMapper {
      * This function uses to handle the field datatype has TypeReference(ex: Record or Enum).
      */
     private Schema handleTypeReference(Map<String, Schema> schema, TypeReferenceTypeSymbol typeReferenceSymbol,
-                                       Schema property) {
+                                       Schema property, boolean isCyclicRecord) {
         if (typeReferenceSymbol.definition().kind() == SymbolKind.ENUM) {
             EnumSymbol enumSymbol = (EnumSymbol) typeReferenceSymbol.definition();
             property = mapEnumValues(enumSymbol);
         } else {
             property.set$ref(typeReferenceSymbol.getName().orElseThrow().trim());
-            createComponentSchema(schema, typeReferenceSymbol);
+            if (!isCyclicRecord) {
+                createComponentSchema(schema, typeReferenceSymbol);
+            }
         }
         return property;
     }
@@ -233,7 +236,7 @@ public class OpenAPIComponentMapper {
      *     };
      * </pre>
      */
-    private Schema handleUnionType(UnionTypeSymbol unionType, Schema property) {
+    private Schema handleUnionType(UnionTypeSymbol unionType, Schema property, String parentComponentName) {
         List<TypeSymbol> unionTypes = unionType.userSpecifiedMemberTypes();
         List<Schema> properties = new ArrayList<>();
         boolean nullable = false;
@@ -243,7 +246,8 @@ public class OpenAPIComponentMapper {
             } else if (union.typeKind() == TypeDescKind.TYPE_REFERENCE) {
                 property = ConverterCommonUtils.getOpenApiSchema(union.typeKind().getName().trim());
                 TypeReferenceTypeSymbol typeReferenceTypeSymbol = (TypeReferenceTypeSymbol) union;
-                property = handleTypeReference(this.components.getSchemas(), typeReferenceTypeSymbol, property);
+                property = handleTypeReference(this.components.getSchemas(), typeReferenceTypeSymbol, property,
+                        isSameRecord(parentComponentName, typeReferenceTypeSymbol));
                 properties.add(property);
                 // TODO: uncomment after fixing ballerina lang union type handling issue
 //            } else if (union.typeKind() == TypeDescKind.UNION) {
@@ -259,6 +263,11 @@ public class OpenAPIComponentMapper {
             property.setNullable(true);
         }
         return property;
+    }
+
+    private boolean isSameRecord(String parentComponentName, TypeReferenceTypeSymbol typeReferenceTypeSymbol) {
+        return typeReferenceTypeSymbol.getName().isPresent() &&
+                parentComponentName.equals(typeReferenceTypeSymbol.getName().get().trim());
     }
 
     /**

--- a/openapi-cli/src/test/java/io/ballerina/openapi/generators/openapi/RecordTests.java
+++ b/openapi-cli/src/test/java/io/ballerina/openapi/generators/openapi/RecordTests.java
@@ -32,8 +32,7 @@ import java.nio.file.Paths;
  * This test class for the covering the unit tests for record scenarios.
  */
 public class RecordTests {
-    private static final Path RES_DIR =
-            Paths.get("src/test/resources/ballerina-to-openapi").toAbsolutePath();
+    private static final Path RES_DIR = Paths.get("src/test/resources/ballerina-to-openapi").toAbsolutePath();
     private Path tempDir;
 
     @BeforeMethod
@@ -87,6 +86,12 @@ public class RecordTests {
     public void testUnionTypeWithNullable() throws OpenApiConverterException, IOException {
         Path ballerinaFilePath = RES_DIR.resolve("record/union_with_nullable.bal");
         TestUtils.compareWithGeneratedFile(ballerinaFilePath, "record/union.yaml");
+    }
+
+    @Test(description = "When the record field has cyclic record type")
+    public void testCyclicRecord() throws IOException {
+        Path ballerinaFilePath = RES_DIR.resolve("record/cyclic_record.bal");
+        TestUtils.compareWithGeneratedFile(ballerinaFilePath, "record/cyclic_record.yaml");
     }
 
     @AfterMethod

--- a/openapi-cli/src/test/resources/ballerina-to-openapi/expected_gen/record/cyclic_record.yaml
+++ b/openapi-cli/src/test/resources/ballerina-to-openapi/expected_gen/record/cyclic_record.yaml
@@ -1,0 +1,62 @@
+openapi: 3.0.1
+info:
+  title: PayloadV
+  version: 0.0.0
+servers:
+  - url: "{server}:{port}/payloadV"
+    variables:
+      server:
+        default: petstore.swagger.io
+      port:
+        default: "443"
+paths:
+  /pet:
+    post:
+      operationId: operation_post_/pet
+      responses:
+        "200":
+          description: Ok
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ResponseError'
+        "202":
+          description: Accepted
+  /pet02:
+    post:
+      operationId: operation_post_/pet02
+      responses:
+        "200":
+          description: Ok
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ResponseError02'
+        "202":
+          description: Accepted
+components:
+  schemas:
+    ResponseError:
+      required:
+        - id
+      type: object
+      properties:
+        id:
+          type: integer
+          format: int32
+          nullable: true
+        resError:
+          $ref: '#/components/schemas/ResponseError'
+    ResponseError02:
+      required:
+        - id
+      type: object
+      properties:
+        id:
+          type: integer
+          format: int32
+          nullable: true
+        resError:
+          oneOf:
+            - $ref: '#/components/schemas/ResponseError02'
+            - type: string

--- a/openapi-cli/src/test/resources/ballerina-to-openapi/record/cyclic_record.bal
+++ b/openapi-cli/src/test/resources/ballerina-to-openapi/record/cyclic_record.bal
@@ -1,0 +1,24 @@
+import ballerina/http;
+
+type ResponseError record {|
+    int? id;
+    ResponseError resError?;
+|};
+
+type ResponseError02 record {|
+    int? id;
+    ResponseError02|string resError?;
+|};
+
+listener http:Listener ep0 = new (443, config = {host: "petstore.swagger.io"});
+
+service /payloadV on ep0 {
+    resource function post pet() returns ResponseError|http:Accepted {
+        http:Accepted accept = {body: ()};
+        return accept;
+    }
+    resource function post pet02() returns ResponseError02|http:Accepted {
+        http:Accepted accept = {body: ()};
+        return accept;
+    }
+}

